File: primitives/transform.ts
import { Primitive } from "./primitive.interface";
import { Quaternion } from "./quaternion";
import { Vector3 } from "./vector3";

export class Transform implements Primitive {
	constructor (
		public position: Vector3 = Vector3.ZERO,
		public rotation: Quaternion = Quaternion.IDENTITY
	) {}

	move(delta: Vector3) {
		this.position = this.position.add(delta);
	}

	rotate(delta: Vector3 | Quaternion) {
		if (delta instanceof Vector3) {
			delta = Quaternion.fromEuler(delta);
		}

		this.rotation = this.rotation.multiply(delta).normalize();
	}

	clone(): Transform {
		return new Transform(this.position.clone(), this.rotation.clone());
	}

	toString(): string {
		return `Transform[${this.position.toString()}, ${this.rotation.toString()}]`;
	}
}

File: src/primitives/transform.rs
use std::fmt::{Display, Formatter, Result};
use crate::{Vector3, primitives::Quaternion};

#[derive(Debug, Clone, PartialEq)]
pub struct Transform {
	pub position: Vector3,
	pub rotation: Quaternion
}

impl Transform {
	pub fn new(position: Option<Vector3>, rotation: Option<Quaternion>) -> Self {
		Self {
			position: position.unwrap_or(Vector3::ZERO),
			rotation: rotation.unwrap_or(Quaternion::IDENTITY),
		}
	}
}

impl Display for Transform {
	fn fmt(&self, f: &mut Formatter<'_>) -> Result {
		write!(f, "Transform[{}, {}]", self.position, self.rotation)
	}
}
